### ISSUE TITLE
`authMiddleware` builds `redirect_url` using `localhost:3000` when NextJS runs inside Docker container

### ISSUE THREAD

**Issue by krzysztof-pyrzyk-ah** [clerkinc/javascript#1338]
- [x] Review the documentation: https://clerk.com/docs
- [x] Search for existing issues: https://github.com/clerkinc/javascript/issues
- [x] Go through package changelog files.
- [ ] Provide the Frontend API key from your application dashboard.

## Package + Version

- [x] `@clerk/nextjs`

### Dependencies + versions

```json
{  
  "dependencies": {
    "@clerk/nextjs": "^4.20.0",
    "next": "13.4.4",
    "react": "18.2.0"
  }
}

```

## Description
Steps to reproduce:
1. Setup repo based on NextJS with Docker [example](https://github.com/vercel/next.js/tree/v13.4.4/examples/with-docker).
2. Install `@clerk/nextjs` and setup `authMiddleware` according to [docs](https://clerk.com/docs/nextjs/middleware).
4. Build the docker image: `docker build -t nextjs-clerk .`
5. Run the image on a different port than 3000: `docker run --rm -p 8000:3000 --env-file ./.env.local nextjs-clerk`.
6. Open a page protected by `authMiddleware`.

Result:
- Sign in url equals `/sign-in?redirect_url=http%3A%2F%2Flocalhost%3A3000`

Expected result:
- Sing in url equals `/sign-in?redirect_url=http%3A%2F%2Flocalhost%3A8000`

Possible solution:
- allow `redirect_url` domain to be configured via env variables

**Comment by dimkl** [comment]
Hello @krzysztof-pyrzyk-ah 👋 
Thank you for opening this issue. I have investigated it and it seems that the `req.nextUrl` or the `req.url` in the NextJS application is not populated correctly in this case.
I have created [this Post](https://discord.com/channels/752553802359505017/1117772771892666510) in the NextJS discord channel to request some help in order to pinpoint the issue.
I will investigate it more and let you know about possible workarounds or if there is a fix.

**Comment by Thisisjuke** [comment]
Hey guys, 
Do you think there is a way to handle a working redirection in a specific case using: `return NextResponse.redirect('/admin')` ? It will be a workaround to avoid exiting the application to a `localhost` url. 

Can you share an `afterAuth` snippet that just avoid to exit the app ? 
Thx :pray: !

**Comment by vbrantner** [comment]
@Thisisjuke One working working workaround, when the cache is cleared, set the sign-path that is defined by clerk. 
```
...
afterAuth(auth, req, evt) {
    // handle users who aren't authenticated
    if (!auth.userId && !auth.isPublicRoute) {
      return NextResponse.redirect("https://accounts.<domain name>/sign-in");
    }
  }
 ...
 ```
After a successful login, the it will redirect to the path specified in the clerk dashboard, since the https://accounts./<domain name>/sign-in is a separate website.

**Comment by Thisisjuke** [comment]
@vbrantner thx for the workaround ! I have a good enough solution atm that can be deployed in pre-production without critical problems.
I now have this snippet in my middleware:

 ```ts
     afterAuth(auth, req, evt) {
        if (!auth.userId && !auth.isPublicRoute) {
            return process.env.NODE_ENV === 'production' ? NextResponse.redirect('https://accounts.<domain>/sign-in') : redirectToSignIn({ returnBackUrl: req.url })
        }
    },
 ```
 
 I am waiting for a better way to handle it but I am not stuck anymore ! :v:

**Comment by mbrookson** [comment]
I'm just been experiencing the same issue when using custom pages for sign-in and sign-up rather than using Clerk hosted pages.

I've got a Next.js app hosted on Railway. In the hosted container I ended up with failing tRPC POST requests (GET seems to work okay though 🤷‍♂️) which were trying to redirect to `http://railway:<PORT>/auth/sign-in`.

I was configuring the URLs with environment variables. Previously I had them set as follows, which works locally and on Vercel:

```sh
NEXT_PUBLIC_CLERK_SIGN_IN_URL=/auth/sign-in
NEXT_PUBLIC_CLERK_SIGN_UP_URL=/auth/sign-up
```

Thanks to the inspiration above, to fix the issue for me I just updated those environment variables to be the fully qualified URL and it works fine now! 🎉 

```sh
NEXT_PUBLIC_CLERK_SIGN_IN_URL=https://my.domain.com/auth/sign-in
NEXT_PUBLIC_CLERK_SIGN_UP_URL=https://my.domain.com/auth/sign-up
```

**Comment by Thisisjuke** [comment]
Thx @mbrookson !

So you didn't change anything inside the `middleware.ts` ? 
I may change to your solution and let my CI/CD do the the trick with env variables. 

Can you please share your `afterAuth` if you have one ? 🙇‍♂️

**Comment by mbrookson** [comment]
Nope, no changes to `middleware.ts` required! I don't have an `afterAuth` function in my middleware 🙂

**Comment by dimkl** [comment]
@mbrookson thank you for the comment but this is not the same issue as the one described.
The original issue is related to the `redirect_url` query param being populated by the `authMiddleware` having hostname `localhost:3000`( eg `https://domain/auth/sign-in?redirect_url=http%3A%2F%2Flocalhost%3A3000%2Fprotected_page`. 
The issue you are describing is related to the hostname of the redirect itself (eg `http://railway:<PORT>/auth/sign-in`).
 I will try to investigate your case separately.

**Comment by mbrookson** [comment]
@dimkl  Ah yep you're right, my bad. Similar but different issue!

**Comment by Thisisjuke** [comment]
Yes, it's not directly linked but the issue is that right now we don't have a way to display a correct login form in production in a dockerized application:

- Using the Clerk documentation example: `redirect_url` is wrong.
- Using the `afterAuth` + `redirectToSignIn`: we are login from `account.<domain>.com` and not directly from inside the application page & layout. 
- Using `NEXT_PUBLIC_CLERK_SIGN_IN_URL` and `NEXT_PUBLIC_CLERK_SIGN_UP_URL`: side effects on my side with script login or things relative to the JS build (`TypeError: Missing parameter name at 5` from `vendors_a74c96_4.47.2.js:2 `). The SignIn form is not displayed.

I understood from your previous message that the original issue is coming from NextJS:
> `req.nextUrl` or the `req.url` in the NextJS application is not populated correctly in this case.

Is there a quick fix to have a working production Sign-in form inside self-hosted docker (not Vercel) ? With both correct `redirect_url` (or cheated redirection) that doesn't exit the application (localhost:3000) & without redirecting to `account.<domain>.com` ?

Thx in advance :pray:

**Comment by nikosdouvlis** [comment]
Hello everyone :)
I wanted to let you know that we're still iterating on possible solutions for this issue. Unfortunately, this is a long-standing "issue" and it's highly unlikely that the NextJS team is going to provide a solution from their end soon.

We have a potential **experimental** solution that we plan to release once testing is done behind an ENV flag for now. If you want to try the solution, you can follow these steps: 

- Install the version containing the workaround: `npm i @clerk/nextjs@4.21.9-snapshot.56dc3e3`
- Add the `CLERK_USE_X_FWD_HEADERS=true` in your ENV (`.env`, `.env.local`, `.env.prod` depending on the service you use and your setup)
- Redeploy the app and test

This version also adds an `experimental_clerkUrl` prop on the `req` object which is a clone of `req.nextUrl` with the correct URL set:
![image](https://github.com/clerkinc/javascript/assets/1811063/be01166b-8229-451b-864a-aec42064ab6e)


It'd be great if you could give it a try and let us know whether this worked or not for you :)

PS: This **should** work for all platforms that host Nextjs apps behind a proxy/ inside a container, like Netlify, Fly.io, AWS, Railway, dockerized containers etc but we haven't been able to test all of them yet.

PS2: More context about the Nextjs issue: 
- https://github.com/vercel/next.js/issues/37536
- https://github.com/vercel/next.js/issues/38724

**Comment by emmems** [comment]
> Hello everyone :) I wanted to let you know that we're still iterating on possible solutions for this issue. Unfortunately, this is a long-standing "issue" and it's highly unlikely that the NextJS team is going to provide a solution from their end soon.
> 
> We have a potential **experimental** solution that we plan to release once testing is done behind an ENV flag for now. If you want to try the solution, you can follow these steps:
> 
> * Install the version containing the workaround: `npm i @clerk/nextjs@4.21.9-snapshot.56dc3e3`
> * Add the `CLERK_USE_X_FWD_HEADERS=true` in your ENV (`.env`, `.env.local`, `.env.prod` depending on the service you use and your setup)
> * Redeploy the app and test
> 
> This version also adds an `experimental_clerkUrl` prop on the `req` object which is a clone of `req.nextUrl` with the correct URL set: ![image](https://user-images.githubusercontent.com/1811063/247284956-be01166b-8229-451b-864a-aec42064ab6e.png)
> 
> It'd be great if you could give it a try and let us know whether this worked or not for you.
> 
> PS: This **should** work for all platforms that host Nextjs apps behind a proxy/ inside a container, like Netlify, Fly.io, AWS, Railway, dockerized containers etc but we haven't been able to test all of them yet.
> 
> PS2: More context about the Nextjs issue:
> 
> * [Hostname of request object in middleware is always 'localhost' vercel/next.js#37536](https://github.com/vercel/next.js/issues/37536)
> * [Middleware behind nginx returning wrong url/origin/host vercel/next.js#38724](https://github.com/vercel/next.js/issues/38724)

Sadly but setup:

Dokku + NextJS 13 

does NOT work :/

More details:
still recognises as http://localhost:3000

buuuut those headers are provided by nginx:
url: 'http://localhost:3000/app',

headers:
x-forwarded-for: '162.158.102.32,::ffff:172.17.0.1',
x-forwarded-host: 'this-is-my-real-domain.example.com',
x-forwarded-port: '80,80',
x-forwarded-proto: 'http,http',
host: 'this-is-my-real-domain.example.com',

**Comment by dimkl** [comment]
@emmems could you provide more information?
- Have you used `npm i @clerk/nextjs@4.21.9-snapshot.56dc3e3` ?
- Did you set `CLERK_USE_X_FWD_HEADERS="true"` environment variable ?
- Do you use any custom beforeAuth / afterAuth ?
- Any other related information ?

I will try to setup dokku and check this out.

**Comment by dimkl** [comment]
@krzysztof-pyrzyk-ah i have tested the `@clerk/nextjs@4.21.9-snapshot.56dc3e3` version with `CLERK_USE_X_FWD_HEADERS="true"` environment variable in testing [repo](https://github.com/dimkl/aws-clerk-nextjs/commit/28df1119425cfb2a96ce7ffe6858f68539b5081e) and it seems that the issue is fixed.
Could you verify that.

**Comment by nikosdouvlis** [comment]
@emmems The `x-forwarded-host` value you shared seems correct so the fix above *should* work for your setup. Can you please provide a sample repo we can take a look at?

**Comment by emmems** [comment]
So my lovely highly responsive friends :) I have fixed that by adding those lines in middleware.ts:

```
afterAuth(auth, req, evt) {
    if (
      !auth.userId &&
      !auth.isPublicRoute &&
      process.env.NODE_ENV == "production"
    ) {
      return NextResponse.redirect(
        `https://{login_domain}/sign-in?redirect_url=${encodeURIComponent(
          "https://{dashboard_domain}/welcome"
        )}`
      );
    }
  },
```

I don't have any example repo for this at least for now :/ 

In case of `CLERK_USE_X_FWD_HEADERS="true"` this was added but it wasn't working sadly :/ 

If I will have more time then I will prepare some examples but I'm not promising anything.

**Comment by Thisisjuke** [comment]
Hi guys, thx for all information above!
@nikosdouvlis coming with feedback about `CLERK_USE_X_FWD_HEADERS=true`.

I am using the version:
```js
"@clerk/nextjs": "^4.21.10",
```
My provider is: https://www.scaleway.com/en/ : we have a private VM with a docker compose running on it.

It's not working properly but it's doing better ! Right now, my url looks like:
```
https://<domain-name>.fr:80/login-in?redirect_url=https%3A%2F%2F<domain-name>.fr%3A80%2Fadmin
```

As you can see, there is a trailing `:80` in my url. It's the right port but right now I don't need it in production. It causes this "SSL" error for the user :
![image](https://github.com/clerkinc/javascript/assets/23522065/e2666a48-5e8a-43b8-bed0-f0a2ff5d52e0)

If I log it like this: 
```js
beforeAuth: (req) => {
    console.log(req.experimental_clerkUrl)
},
```

I have the following object:
```js
{
  href: 'https://<domain-name>.fr:80/',
  origin: 'https://<domain-name>.fr:80',
  protocol: 'https:',
  username: '',
  password: '',
  host: '<domain-name>.fr:80',
  hostname: '<domain-name>.fr',
  port: '80',
  pathname: '/',
  search: '',
  searchParams: <ref *1> URLSearchParams {
  [Symbol(query)]: [],
  [Symbol(context)]: URL {
  [Symbol(context)]: URLContext {
  href: 'https://<domain-name>.fr:80/',
  origin: 'https://<domain-name>.fr:80',
  protocol: 'https:',
  hostname: '<domain-name>.fr',
  pathname: '/',
  search: '',
  username: '',
  password: '',
  port: '80',
  hash: ''
},

```

Has this been seen or dealt with ?

**Comment by dimkl** [comment]
@Thisisjuke sorry for the inconvenience we are still working on some edge cases. The described behavior seems to be the same as this reported issue: https://github.com/clerkinc/javascript/issues/1413
I have already a PR with a fix but it needs to be tested across multiple platforms. We will ping you back about any progress.

**Comment by krzysztof-pyrzyk-ah** [comment]
I've tested the solution mentionet above, with `@clerk/nextjs@4.21.9-snapshot.56dc3e3` and the `experimental_clerkUrl` and it seems to be working just fine inside Docker containers.

Thank you @nikosdouvlis and @dimkl 🙇‍♂️

**Comment by dimkl** [comment]
Fix released in version [@clerk/nextjs@4.21.11](https://github.com/clerkinc/javascript/releases/tag/%40clerk/nextjs%404.21.11)

PS: We encourage the use `CLERK_TRUST_HOST: "true"` environment variable instead of `CLERK_USE_X_FWD_HEADERS: "true"` since it makes more sense as to what that env does. The `CLERK_USE_X_FWD_HEADERS` will be deprecated in the future.

**Comment by Thisisjuke** [comment]
@dimkl, we are now using [@clerk/nextjs@4.21.11](https://github.com/clerkinc/javascript/releases/tag/%40clerk/nextjs%404.21.11) in both the development and production environments, and everything is working well ! There is no trailing port :80 within our Scaleway container app.
I will reach out again if we encounter any misbehaviors or other issues.
Thank you for your feedback and availability !🙏

**Comment by dimkl** [comment]
We have enabled by default the fix in version [@clerk/nextjs@4.23.0](https://github.com/clerkinc/javascript/releases/tag/%40clerk/nextjs%404.23.0), so there is no need to set any environment variable anymore.

**Comment by clerk-cookie** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
